google/cloud/storage/internal/async/writer_connection_buffered.cc

```diff
@@ -0,0 +1,405 @@
+    }
+    auto const n = persisted_size - buffer_offset_;
+    if (n > static_cast<std::int64_t>(resend_buffer_.size())) {
+      return SetError(std::move(lk),
+                      MakeFastForwardError(UploadId(), buffer_offset_,
+                                           persisted_size, GCP_ERROR_INFO()));
+    }
+    resend_buffer_.RemovePrefix(static_cast<std::size_t>(n));
+    buffer_offset_ = persisted_size;
+    write_offset_ -= static_cast<std::size_t>(n);
+    // If the buffer is small enough, collect all the handlers to notify them.
+    auto const handlers = ClearHandlersIfEmpty(lk);
+    WriteLoop(std::move(lk));
+    // The notifications are deferred until the lock is released, as they might
+    // call back and try to acquire the lock.
+    for (auto const& h : handlers) h->Execute(Status{});
+  }
+
+  void WriteStep(std::unique_lock<std::mutex> lk, absl::Cord payload) {
+    auto const size = payload.size();
+    lk.unlock();
+    (void)impl_->Write(WritePayloadImpl::Make(std::move(payload)))
+        .then([this, size](auto f) { OnWrite(f.get(), size); });
+  }
+
+  void OnWrite(Status result, std::size_t write_size) {
+    if (!result.ok()) return Resume(std::move(result));
+    std::unique_lock<std::mutex> lk(mu_);
+    write_offset_ += write_size;
+    return WriteLoop(std::move(lk));
+  }
+
+  void Resume(Status s) {
+    if (!s.ok() && cancelled_) {
+      return SetError(std::unique_lock<std::mutex>(mu_), std::move(s));
+    }
+    factory_().then([this](auto f) { OnResume(f.get()); });
+  }
+
+  void OnResume(
+      StatusOr<std::unique_ptr<storage_experimental::AsyncWriterConnection>>
+          impl) {
+    std::unique_lock<std::mutex> lk(mu_);
+    if (!impl) return SetError(std::move(lk), std::move(impl).status());
+    impl_ = *std::move(impl);
+    auto state = impl_->PersistedState();
+    if (absl::holds_alternative<storage::ObjectMetadata>(state)) {
+      finalized_.set_value(
+          absl::get<storage::ObjectMetadata>(std::move(state)));
+      auto handlers = ClearHandlers(lk);
+      lk.unlock();
+      for (auto& h : handlers) h->Execute(Status{});
+      return;
+    }
+    OnQuery(std::move(lk), absl::get<std::int64_t>(state));
+  }
+
+  void SetError(std::unique_lock<std::mutex> lk, Status status) {
+    resume_status_ = status;
+    writing_ = false;
+    auto handlers = ClearHandlers(lk);
+    lk.unlock();
+    for (auto& h : handlers) h->Execute(status);
+    finalized_.set_value(std::move(status));
+  }
+
+  // Creates new `impl_` instances when needed.
+  WriterConnectionFactory const factory_;
+
+  // Request a server-side fush if the buffer goes over this threshold.
+  std::size_t const buffer_size_lwm_;
+
+  // Stop sending data if the buffer goes over this threshold. Only
+  // start sending data again if the size goes below buffer_size_lwm_.
+  std::size_t const buffer_size_hwm_;
+
+  // The background threads may change the state of this class
```

Comment:
Add a comment somewhere to indicate what this is protecting like `// ABSL_GUARDED_BY(mu_)`

Seems like it might be protecting a lot of things...

Reply:
Done.  All the remember member variables are protected by that mutex.